File: src/2337_Move_Pieces_to_Obtain_a_String.py
# 2337. Move Pieces to Obtain a String
# https://leetcode.cn/problems/move-pieces-to-obtain-a-string/description/


# The condition `if i != j and (c == 'L') == (i < j):` contains two parts, 
# and understanding both is crucial for grasping the logic of the algorithm:

# 1. `i != j`: This checks whether the current character from the `start` 
# string and the corresponding character from the `target` string are in different positions. 
# If `i == j`, it means the current character hasn't moved, 
# so there's no need to check any further for this character since it's already in the correct position in the target string. 

# 2. `(c == 'L') == (i < j)`: This condition is used to determine the legality of the move. 
# If the character is 'L', it means it can only move to the left. Hence, its starting position (`i`) 
# should be greater than its ending position (`j`). If the character is 'R', it can only move to the right, 
# so its starting position should be less than its ending position. Thus, the condition checks for the legality 
# of the move based on the character and its positions in both strings.

# Putting the two conditions together:

# - If `i == j`, the character hasn't moved, and we continue with the next character.
# - If `i != j`, the character has moved, and then we need to check if the move was legal. The legality is checked by the second part of the condition.

# So, `i != j` acts as a filter to only inspect moves where the character has actually changed its position. If we didn't include this condition, it would be possible to flag legal non-moves (characters that haven't moved) as illegal.
class Solution:
    def canChange(self, start: str, target: str) -> bool:
        if start.replace('_', '') != target.replace('_', ''):
            return False

        j = 0
        for i, c in enumerate(start):
            if c == '_':
                continue
            
            while target[j] == '_':
                j += 1
            
            if i != j and (c == 'L') == (i < j):
                return False
            
            j += 1
            
        return True
            

File: src/2337_Move_Pieces_to_Obtain_a_String.rs
// 2337_Move_Pieces_to_Obtain_a_String
// https://leetcode.cn/problems/move-pieces-to-obtain-a-string/

// You are given two strings start and target, both of length n. Each string consists only of the characters 'L', 'R', and '_' where:

// The characters 'L' and 'R' represent pieces, where a piece 'L' can move to the left only if there is a blank space directly to its left, and a piece 'R' can move to the right only if there is a blank space directly to its right.
// The character '_' represents a blank space that can be occupied by any of the 'L' or 'R' pieces.
// Return true if it is possible to obtain the string target by moving the pieces of the string start any number of times. Otherwise, return false.

// 超时

/// The target.chars().nth(j) call is the cause of the timeout. 
/// In Rust, strings are UTF-8 encoded, and accessing a character by its index using nth() is an O(n) operation 
/// since it has to iterate over the bytes to determine the nth character due to variable-length encoding. 
/// Therefore, within the loop, repeatedly calling nth() can cause the algorithm to degrade into O(n^2) complexity.
impl Solution {
    pub fn can_change(start: String, target: String) -> bool {
        if start.replace("_", "") != target.replace("_", "") {
            return false;
        }

        let mut j = 0;
        for (i, c) in start.chars().enumerate() {
            if c == '_' {
                continue;
            }
            while target.chars().nth(j).unwrap_or(' ') == '_' {
                j += 1;
            }
            if i != j && (c == 'L') == (i < j) {
                return false;
            }
            j += 1;
        }
        true
    }
}


impl Solution {
    pub fn can_change(start: String, target: String) -> bool {
        if start.replace("_", "") != target.replace("_", "") {
            return false;
        }

        let mut j = 0;
        let mut target_chars = target.chars().peekable();
        for (i, c) in start.chars().enumerate() {
            if c == '_' {
                continue;
            }
            
            while target_chars.peek() == Some(&'_') {
                target_chars.next();
                j += 1;
            }
            
            if i != j && (c == 'L') == (i < j) {
                return false;
            }
            target_chars.next();
            j += 1;
        }
        
        true
    }
}